### PR TITLE
[@types/underscore] Fix #36308 - underscore has bad chain typings fix only for function 'map' (#36319)

### PR DESCRIPTION
**Pull request by Radullus**
Fix typing in 'underscore' for function 'map'

Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: https://underscorejs.org/#map
https://underscorejs.org/#chain
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by Radullus** [comment]
@jgonggrijp @regevbr @renjfk  Can you please review?

**Comment by jgonggrijp** [comment]
I defer this one to @regevbr because he is currently working on a complete fix for the chain typings. I'm not sure whether the fix in the current PR will combine well with his work, but I think the unittest would be a welcome addition in any case.

**Review by regevbr — CHANGES_REQUESTED** [review]
@Radullus Thanks for the PR!

The `T` in `_Chain<T,V>` represents the type of the object and `V` represents the object.  
So for example, `string[]` should be `_Chain<string,string[]>` and `_.Dictionary<number>` should be `_Chain<number, _.Dictionary<number>` 

The lines you changed made your test pass, but they would cause other tests, that should be there but are not :), to fail.

Please see the following suggestion to fix the issue:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d200340ecb521b3856f8bbf6e5b61a33182f9363/types/underscore/index.d.ts#L5103

should be changed to 
```ts
 map<TResult>(iterator: _.ListIterator<T, TResult>, context?: any): _Chain<TResult, TResult[]>; 
```

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d200340ecb521b3856f8bbf6e5b61a33182f9363/types/underscore/index.d.ts#L5115

should be changed to 
```ts
 map<TResult>(iterator: _.ObjectIterator<T, TResult>, context?: any): _Chain<TResult, _.Dictionary<TResult>>;
```

Can you please try it (after reverting your change) and see if the test passes? Also I agree with @jgonggrijp that more tests are always welcome, so can you please also add a test for map operation on an object?

File: types/underscore/index.d.ts [hunk 1]
```diff
@@ -5094,25 +5094,25 @@ declare module _ {
         * Wrapped type `any[]`.
         * @see _.map
         **/
-        map<TArray>(iterator: _.ListIterator<T, TArray[]>, context?: any): _ChainOfArrays<TArray>;
+        map<TArray>(iterator: _.ListIterator<T, TArray>, context?: any): _ChainOfArrays<TArray>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: this should be reverted as current behavior is ok

File: types/underscore/index.d.ts [hunk 2]
```diff
@@ -5094,25 +5094,25 @@ declare module _ {
...
         **/
-        map<TResult>(iterator: _.ListIterator<T, TResult>, context?: any): _Chain<TResult>;
+        map<TResult>(iterator: _.ListIterator<T, TResult>, context?: any):  _Chain<TResult, TResult[]>;
 
         /**
         * Wrapped type `any[]`.
         * @see _.map
         **/
-        map<TArray>(iterator: _.ObjectIterator<T, TArray[]>, context?: any): _ChainOfArrays<TArray>;
+        map<TArray>(iterator: _.ObjectIterator<T, TArray>, context?: any): _ChainOfArrays<TArray>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: this should be reverted as current behavior is ok

**Comment by Radullus** [comment]
- I added more test, for mapping 'Dictionary'.
 - I apply charges suggested for Line 5103 in line DefinitelyTyped/types/underscore/index.d.ts d200340
 - I modified Line 5115 DefinitelyTyped/types/underscore/index.d.ts d200340. 
Map function for 'chain' object don't return 'Dictionary' but return 'Array'.  You can see some of my tested https://jsfiddle.net/Radullus/fsmj2rxa/37/

File: types/underscore/underscore-tests.ts [hunk 3]
```diff
@@ -209,6 +209,28 @@ _(stooges)
...
+    .map(p => p.name)
+    .value();
+
+var usersData: _.Dictionary<{ age: number; name: string }> = {
+    'user id': { name: 'moe', age: 40 },
+    'other user Id': { name: 'larry', age: 50 },
+    'fake id': { name: 'curly', age: 60 },
+};
+
+let youngPeopleId: string[] = _.chain(usersData)
```
Comment (regevbr) [inline types/underscore/underscore-tests.ts]: This test should actually fail, as performing filter on a a Dictionary<T> returns T[], thus performing map on it will need the following iterator:
```
(value: T, index: number, list: T[]): TResult
```
The reason this test passes now is because of missing signature of filter for _Chain which should be:
```ts
filter(iterator: _.ObjectIterator<T, boolean>, context?: any): _Chain<T, T[]>;
```
But fixing it will not even help as the entire _Chain interface is malformed :-( and I'm working on a wide fix for it.
I would rather you only add tests that use the fixed map method

**Comment by regevbr** [comment]
Thanks @Radullus!
The _Chain interface is completely messed up currently :(
I added some comments for you to review

**Comment by Radullus** [comment]
Hi @regevbr . I have applied the changes you had suggested.

**Comment by regevbr** [comment]
I added new comment for you to review

File: types/underscore/underscore-tests.ts (hunk 3 above)
Comment (Radullus) [reply types/underscore/underscore-tests.ts]: You have right. I will stick in this test to 'map' function. I interface definition I was trying to test.
And, I will add some test for 'filter' separate.

File: types/underscore/index.d.ts (hunk 2 above)
Comment (Radullus) [reply types/underscore/index.d.ts]: done.
I have add test for this definition of 'map', but I have to comment part of the 'testing' witch check output because it will cause the test to fail.

File: types/underscore/index.d.ts (hunk 1 above)
Comment (Radullus) [reply types/underscore/index.d.ts]: done.
I have add test for this definition of 'map', but I have to comment part of the 'testing' witch check output because it will cause the test to fail.

File: types/underscore/underscore-tests.ts [hunk 4]
```diff
@@ -209,6 +209,46 @@ _(stooges)
...
+    .value();
+
+let usersTable: { age: number; name: string; id: string }[] = _.chain(usersData)
+    .map<{ age: number; name: string; id: string }>((p, k: string) => {
+        return { id: k, ...p };
+    })
+    .value();
+
+// Test map function with _ChainOfArrays<>
+let usersTable_2 /*: { age: number; name: string; id: string }[][]*/ = _.chain(usersData)
```
Comment (regevbr) [inline types/underscore/underscore-tests.ts]: why did you remove the type here?
It is best to use the flatten() method here and then get the value

File: types/underscore/underscore-tests.ts [hunk 5]
```diff
@@ -209,6 +209,46 @@ _(stooges)
...
+    .map<{ age: number; name: string; id: string }>((p, k: string) => {
+        return [{ id: k, ...p }];
+    })
+    .value();
+
+let usersTable_3 /*: { score: number; fullName: string; login: string }[][]*/ = _.chain(usersTable)
```
Comment (regevbr) [inline types/underscore/underscore-tests.ts]: same here - why did you remove the type?

File: types/underscore/underscore-tests.ts (hunk 4 above)
Comment (Radullus) [reply types/underscore/underscore-tests.ts]: because the type by the underscore should be like this one in the comment, But with current typescript definition, There is missing one level of array nest. 
`{ age: number; name: string; id: string }[][]  // type return by underscore
{ age: number; name: string; id: string }[]    // type with is calculate base of current definitions
`
It would require more changes in the underscore typescript definition. And I'm not certain what should be changed?
`interface _ChainOfArrays<T> extends _Chain<T[]> `
to 
`interface _ChainOfArrays<T> extends _Chain<T[], T[][]> `
it will not match with data return from function 'groupBy'

Comment (regevbr) [reply types/underscore/underscore-tests.ts]: You are correct. Lets leave it be for now

**Review by regevbr — APPROVED** [review]
